File: src/main/java/com/demo/t_web/program/memo/model/AskMaker.java
package com.demo.t_web.program.memo.model;

import lombok.Builder;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

/**
 * <pre>
 * com.demo.t_web.program.memo.model.AskVo
 *   - AskVo.java
 * </pre>
 *
 * @author : tarr4h
 * @className : AskVo
 * @description :
 * @date : 12/27/24
 */
@Setter
@Builder
public class AskMaker {

    private String question;

    private String ask;

    private List<Memo> memos;

    private int askType;

    private boolean referenceYn;

    private String conversationId;

    public String getConversationId() {
        return this.conversationId;
    }

    public boolean isReferenceYn() {
        return this.referenceYn;
    }

    public String getAsk(){
        StringBuilder str = new StringBuilder();
        str.append("\n[\n");

        memos.forEach(memo -> {
            str.append("------------------\n")
                    .append("title : ")
                    .append(memo.getTitle())
                    .append("\n")
                    .append("content : \n")
                    .append(memo.getContent())
                    .append("\n")
                    .append("------------------\n");
        });

        str.append("]\n");
        if(this.askType == 2){
            str.append("question : ")
                .append(this.question);
        }
        return str.toString();
    }

    public void addMemo(Memo memo){
        if(this.memos == null) {
            this.memos = new ArrayList<>();
        }
        this.memos.add(memo);
    }

    public void addMemoList(List<Memo> memoList){
        if(this.memos == null) this.memos = new ArrayList<>();
        memos.addAll(memoList);
    }

    public String getPrompt(){
        return isReferenceYn() ? prompt2 : prompt1;
    }

    private final String prompt1 = """
            이전 답변을 참고해서 짧고 간결하게 대답.
            """;
    private final String prompt2 = """
            Make responses quickly with short and clear sentences. \s
            Analyze the content inside the brackets [ ] and answer the question based only on the given information. \s
            Do not infer or mix relationships such as "competitor product" and "counterpart product" unless explicitly stated in the document. \s
            For terms with similar meanings within the same context (e.g., "counterpart product" and "comparable product"), infer only if the relationship is explicitly stated or directly implied in the document. \s
            Do not modify or paraphrase any words inside the brackets. \s
            When multiple products are listed as counterparts to a single product, treat each listed product as a counterpart to the main product only, not to the other listed products. \s
            Each product relationship is specific to the context provided and should not extend beyond the explicitly mentioned information. \s
            Do not assume any implied relationship between listed products unless explicitly stated in the document. \s
            The relationship between products must be understood as directly related to the subject (e.g., "A's counterpart product is B"). \s
            Provide only the answer to the question in Korean, without additional explanations or context. \s
            Use only the information inside the brackets to form the answer, and do not include unrelated or additional responses. \s
            If there is a document stating "A's counterpart products are B, C" then B and C can also be considered A's counterpart. For such inference-based answers, analyze relationships to respond appropriately. \s
            Ensure the response is accurate, even if analysis is required, but keep it short and clear.
            """;
}

File: src/main/java/com/demo/t_web/program/memo/model/AskMaker.kt
package com.demo.t_web.program.memo.model

data class AskMaker (
    var question : String,
    var memos : MutableList<String> = mutableListOf(),
    var askType : Int,
    var referenceYn : Boolean,
    var conversationId : String
){
    fun getAsk() : String {
        val str : StringBuilder = StringBuilder()
        str.append("\n[\n")

        memos.forEach {
            memo -> str.append("------------------\n")
                .append(memo)
                .append("\n")
        }

        str.append("]\n")
        if(askType == 2){
            str.append("question : ")
                .append(question)
        }

        return str.toString()
    }

    fun addMemoList(memoList : List<String>) {
        memos.addAll(memoList)
    }

    fun getPrompt() : String {
        return if(referenceYn) prompt2 else prompt1
    }

    private val prompt1 = """
            이전 답변을 참고해서 짧고 간결하게 대답.        
            """.trimIndent();
    private val prompt2 = """
            Make response quickly with short and clear sentence.
            Analyze the content inside the brackets [ ] and answer the question based on the given information only.
            Do not modify or paraphrase any words inside the brackets.
            Provide only the answer to the question without any additional explanations or context.
            The answer must be provided in Korean.
            Answer should be correct with content inside the brackets mentions.
            Only use the information from the brackets to form the answer, and do not include any unrelated or additional responses.
            Response must be short and clear.
            Define relationships:
            - '<->' means "counterproductive relationship."
            - Analyze ONLY the part of the sentence containing '<->'.
            - Items on the left of '<->' are counterproductive to all items on the right.
            - Items separated by commas on the same side (e.g., B, C) are *not* counterproductive to each other and should NOT be linked in any way.
            
            Relationships:
            A(word) <-> B(word), C(word)
            
            Important:
            - Counterproductive relationships only exist across '<->', NOT within the same side.
            - A is counterproductive to B also A is counterproductive to C.
            - B is counterproductive to A.
            - C is counterproductive to A.
            - B is NOT counterproductive to C.
            - C is NOT counterproductive to B.
            """.trimIndent();
}
//            if [A <-> B \n A, A <-> C] then A is counterproduct of B and C, B is not counterproduct of C
